client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transaction/XaModeTransactionTable.java

```diff
@@ -0,0 +1,370 @@
+               default:
+                  //other error codes
+                  unknown = true;
+                  unknownErrorCode = xa_code;
+                  break;
+            }
+         }
+
+         if (!hasCommit && !hasRollback) {
+            //nothing committed neither rolled back
+            if (alreadyForgot) {
+               //no tx
+               throw new XAException(XAException.XAER_NOTA);
+            } else if (unknown) {
+               //unknown. we throw what we received
+               throw new XAException(unknownErrorCode);
+            }
+         } else if (hasCommit && hasRollback) {
+            //we have something committed and something rolled back
+            throw new XAException(XAException.XA_HEURMIX);
+         } else if (hasRollback) {
+            //rollbacks only
+            throw new XAException(XAException.XA_HEURRB);
+         }
+         // commits only.
+      }
+
+      private int internalPrepare(boolean onePhaseCommit) throws XAException {
+         boolean readOnly = true;
+         for (TransactionContext<?, ?> ctx : registeredCaches.values()) {
+            switch (ctx.prepareContext(currentXid, onePhaseCommit)) {
+               case XAResource.XA_OK:
+                  readOnly = false;
+                  preparedCaches.add(ctx);
+                  break;
+               case XAResource.XA_RDONLY:
+                  break; //read only tx.
+               case Integer.MIN_VALUE:
+                  //signals a marshaller error of key or value. the server wasn't contacted
+                  throw new XAException(XAException.XA_RBROLLBACK);
+               default:
+                  //any other code we need to rollback
+                  //we may need to send the rollback later
+                  preparedCaches.add(ctx);
+                  throw new XAException(XAException.XA_RBROLLBACK);
+            }
+         }
+         return readOnly ? XAResource.XA_RDONLY : XAResource.XA_OK;
+      }
+
+      private <K, V> TransactionContext<K, V> registerCache(TransactionalRemoteCacheImpl<K, V> txRemoteCache) {
+         if (currentXid == null) {
+            throw new CacheException("XaResource wasn't invoked!");
+         }
+         //noinspection unchecked
+         return (TransactionContext<K, V>) registeredCaches
+               .computeIfAbsent(txRemoteCache.getName(), s -> createTxContext(txRemoteCache));
+      }
+
+      private <K, V> TransactionContext<K, V> createTxContext(TransactionalRemoteCacheImpl<K, V> remoteCache) {
+         if (trace) {
+            log.tracef("Registering remote cache '%s' for transaction xid=%s", remoteCache.getName(), currentXid);
+         }
+         return new TransactionContext<>(remoteCache.keyMarshaller(), remoteCache.valueMarshaller(),
+               remoteCache.getOperationsFactory(), remoteCache.getName());
+      }
+
+      private void cleanup() {
+         registeredTransactions.remove(transaction);
+         registeredCaches.clear();
```

Comment:
Do we need to clear/null the fields? I'd expect that this adapter gets forgotten and gced...

Reply:
I think we do. 
Although we don't have recovery in place, the TM can keep this resource for a while if another resource may need some recovery.
Also, (I need to investigate), when recovery is in place, the instance may reincarnate as another transaction and we need to repopulate the `preparedCaches` and `currentXid`.